llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,224 @@
+
+#define DEBUG_TYPE "loop-iter-count-assumptions"
+
+using namespace llvm;
+
+namespace {
+
+std::string getFunctionAndBlockNames(BasicBlock &BB) {
+  return BB.getParent()->getName().str() + " " + BB.getName().str();
+}
+
+/// Return the Branch Compare Instruction of CurrentLoop if the Loop is well
+/// formed and this pass can process the Predicate
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+
+  if (CurrentLoop.isRotatedForm()) {
+    LLVM_DEBUG(dbgs() << "Loop already in rotated form. Will not add Loop "
+                         "Iteration Count assumptions.\n");
+    return nullptr;
+  }
+
+  /// Check that the loop has a single Exiting Block. If the CurrentLoop
+  /// contains multiple Exiting Block, ExitBB will be a nullptr
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return nullptr;
+
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI)
+    return nullptr;
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return nullptr;
+
+  LLVM_DEBUG(dbgs() << "Condition Found: " << *LoopCmpInstr << "\n");
+
+  // Do not process equal and not-equal Compare Instructions, since they do not
+  // work well with the currently used assumption-construction-method of reusing
+  // the ICmpInst Predicate
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return nullptr;
+  }
+  return LoopCmpInstr;
+}
+
+/// Return the AddRecExpr evaluated at Iteration 0 if an AddRecExpr can be
+/// extracted, otherwise return Op
+Value *expandValueAtZeroIteration(Value *Op, ScalarEvolution &SE,
+                                  SCEVExpander &Expander,
+                                  Instruction *InsertionPoint) {
+  const SCEVAddRecExpr *AddRec =
+      dyn_cast_or_null<SCEVAddRecExpr>(SE.getSCEV(Op));
+  if (!AddRec) {
+    LLVM_DEBUG(dbgs() << "Could not extract AddRecExpr, will reuse " << *Op
+                      << "\n");
+    return Op;
+  }
+
+  return Expander.expandCodeFor(
+      AddRec->evaluateAtIteration(SE.getConstant(APInt(32, 0)), SE),
+      Op->getType(), InsertionPoint);
+}
+
+Value *recursivlyCloneInBB(Value *Op, BasicBlock &BB, ValueToValueMapTy &VMap) {
+  if (isa<PHINode>(Op)) {
+    LLVM_DEBUG(dbgs() << "Phi node: do not clone " << *Op << "\n");
+    // return Value that is in the BB, i.e. in the preheader
+    PHINode *PN = dyn_cast<PHINode>(Op);
+    return PN->getIncomingValueForBlock(&BB);
+  }
+
+  Instruction *I = dyn_cast<Instruction>(Op);
+  if (!I) {
+    LLVM_DEBUG(dbgs() << "No Instruction: do not clone: " << *Op << "\n");
+    return Op;
+  }
+  if (I->isVolatile()) {
+    LLVM_DEBUG(dbgs() << "Volatile Instruction: Aborting! Could not clone "
+                      << *Op << "\n");
+    return nullptr;
```

Comment:
As said before, I think we need to check more than just `volatile`, we cannot just clone any instruction without checking it is loop-invariant. Or at least we need to prove that the instructions that precede it will have no effects on its value.